rfcs/010-per-document-access-control.md

```diff
@@ -0,0 +1,418 @@
+---
+name: Per-Document Access Control
+about: Make the db-per-user pattern obsolete.
+title: 'Per-Document Access Control'
+labels: rfc, discussion, access control, security
+assignees: '@janl'
+
+---
+
+# Introduction
+
+Up until now (version 2.3.1), CouchDB could not serve mutually
+untrusting users accessing the same database. If a user has access to
+one document in a database, they have access to all other documents in
+the database. Some restrictions can be added about writing documents
+(designs docs are db-admin only, validate doc update (VDU) functions
+could restrict write access based on the writing user and/or the target
+document). For the remainder of this document, “db-admin” SHALL include
+server admins as well.
+
+## Abstract
+
+This led to CouchDB developers making use of a pattern called
+db-per-user, where all documents belonging to one user are kept in a
+separate database. This is a decent enough workaround, but has the
+following downsides:
+
+- queries across all databases are not possible. An additional
+  workaround exists where all per-user databases are replicated
+  continuously into a central, admin-only database that can be used for
+  querying the entire data set, but that adds latency and uses
+  significant CPU resources. Successful systems have been built where
+  increased latency could be traded for fewer CPU resources, but
+  overall, this is not an optimal design.
+
+- handling many small databases, say >10000 (depending on hardware) can
+  become a challenge, if most of them are active concurrently. It
+  forces dbs to be set to `q=1`, migrating off `q!=1` requires
+  downtime, 10k bidirectional replications are going to need A LOT of
+  CPU and RAM. sharing documents among two or more users requires the
+  creation of yet more databases.
+
+Per-user document access aims to solve many of the above problems.
```

Comment:
I think this paragraph and the goals are probably all thats needed for an abstract.